### PR TITLE
Admin login and problems_controller authorization

### PR DESCRIPTION
**Pull request by somnath77**
```

1. Deleted commented code in en.yml
2. Added devise_token_auth to user
3. Added authorization to problems_controller ( different for admin & reviewer )
4. Wrote unit tests for the same

```

### CODE REVIEW

File: app/controllers/api/v1/admin/problems_controller.rb [hunk 1]
```diff
@@ -42,9 +46,18 @@ def update
...
 
         def problem_params
+          params['created_by_id'] = current_user.id
+          params['updated_by_id'] = current_user.id
+          params['organization_id'] = current_user.organization_id
           params.permit(:id, :title, :description, :created_by_id, :updated_by_id, :organization_id,
                         :drive_id)
         end
+
+        def user_admin?
```
Comment (rahulknojha) [inline app/controllers/api/v1/admin/problems_controller.rb:56]: Lets discuss this

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,4 +1,6 @@
 # frozen_string_literal: true
 
 class ApplicationController < ActionController::Base
+  include DeviseTokenAuth::Concerns::SetUserByToken
```
Comment (rahulknojha) [inline app/controllers/application_controller.rb:4]: why this is needed here ? when u have already included in api controller

File: spec/request/api/v1/admin/problems_spec.rb [hunk 3]
```diff
@@ -4,80 +4,230 @@
...
+      let(:role) { create(:role) }
+      let(:user) { create(:user) }
+      let(:problem) do
+        create(:problem, created_by_id: user.id,
+                         updated_by_id: user.id, organization: organization)
+      end
+
+      it 'Creates the problem' do
+        @request.env['devise.mapping'] = Devise.mappings[:user]
+        request.headers.merge! user.create_new_auth_token
```
Comment (rahulknojha) [inline spec/request/api/v1/admin/problems_spec.rb:19]: move this to a helper method, Do not repeat yourself

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (somnath77) [reply app/controllers/application_controller.rb]: removed

File: spec/request/api/v1/admin/problems_spec.rb (hunk 3 above)
Comment (somnath77) [reply spec/request/api/v1/admin/problems_spec.rb]: Moved token generation code into helper file

**Review by rahulknojha — APPROVED** [review]

File: app/controllers/api/v1/admin/problems_controller.rb [hunk 4]
```diff
@@ -4,8 +4,12 @@ module Api
...
       class ProblemsController < ApiController
+        before_action :authenticate_user!
+        load_and_authorize_resource
+
         def create
           problem = Problem.new(problem_params)
+
           if problem.save
             render_success(data: { problem: serialize_resource(problem, ProblemSerializer) },
                            message: I18n.t('create.success', model_name: Problem))
```
Comment (varungadde99) [inline app/controllers/api/v1/admin/problems_controller.rb:15]: what is problem doesnt gets saved due to some error.
Please add the render_error part if possible.

Comment (somnath77) [reply app/controllers/api/v1/admin/problems_controller.rb:15]: It is there, please see full file

Comment (somnath77) [reply app/controllers/api/v1/admin/problems_controller.rb:15]: It is there, please see full file

**Review by varungadde99 — APPROVED** [review]

File: app/controllers/api_controller.rb [hunk 5]
```diff
@@ -1,8 +1,14 @@
 # frozen_string_literal: true
 
 class ApiController < ActionController::API
+  include DeviseTokenAuth::Concerns::SetUserByToken
+
   rescue_from ActiveRecord::RecordNotFound, with: :error_render_method
 
+  rescue_from CanCan::AccessDenied do |_exception|
+    render json: { message: I18n.t('reviewer.error') }, status: 403
```
Comment (rahulknojha) [inline app/controllers/api_controller.rb:9]: exception we are handling is for authorization, but message key does not relate at all `reviewer.error`

File: spec/controllers/api/v1/admin/problems_spec.rb [hunk 6]
```diff
@@ -6,78 +6,137 @@
...
+        expect(problem['data']['problem']['title']).to eq('a')
+        expect(response.has_header?('access-token')).to eq(true)
+        expect(response).to have_http_status(:ok)
+      end
 
-      expect(problem['data']['problem']['title']).to eq('a')
-      expect(response).to have_http_status(:ok)
+      it ' creates problem even though not passing organization id,
+        created_by_id, it takes from sessions' do
+        auth_tokens_for_user(user)
```
Comment (rahulknojha) [inline spec/controllers/api/v1/admin/problems_spec.rb:33]: can be moved to before each block if needed in all specs in a context

**Review by rahulknojha — APPROVED** [review]

File: spec/controllers/api/v1/admin/problems_spec.rb (hunk 6 above)
Comment (somnath77) [reply spec/controllers/api/v1/admin/problems_spec.rb:33]: user` is not available on an example group (e.g. a `describe` or `context` block). It is only available from within individual examples (e.g. `it` blocks)

Error occured, leaving it as it is

File: app/controllers/api_controller.rb (hunk 5 above)
Comment (somnath77) [reply app/controllers/api_controller.rb:9]: Changed it to `user.unauthorized`

File: spec/controllers/api/v1/admin/problems_spec.rb (hunk 6 above)
Comment (rahulknojha) [reply spec/controllers/api/v1/admin/problems_spec.rb:33]: ok we will discuss later on, note it down and ask me whenever I am in office we will see together how we can frame this.

**Review by rahulknojha — APPROVED** [review]

File: spec/controllers/api/v1/admin/problems_spec.rb (hunk 6 above)
Comment (somnath77) [reply spec/controllers/api/v1/admin/problems_spec.rb:33]: OK sir

**Pull request merged**